### PR TITLE
Rewrite Go JS core

### PR DESCRIPTION
**Pull request by PlasmaPower**
This PR was co-authored with @rachel-bousfield so it should be reviewed by someone else.

This PR is currently a draft because it doesn't implement stylus-specific functionality like callbacks into Go (and comments out stylus tests for now). That said, I think it'd make sense to review this PR in its current state to make sure this approach makes sense before adding on Stylus callbacks. Everything implemented in this PR should be complete and ready for review.

Broadly speaking, this PR largely combines the go-stub wasm-libraries Go JS implementation and the JIT Go JS implementation into a single go-js library, and switches from an imperative programming style to a declarative style.

Before this PR, methods would be implemented with a match statement against the object ID and the method name:

```rs
    let value = match class {
        UINT8_ARRAY_ID => match args.get(0) {
            Some(JsValue::Number(size)) => DynamicObject::Uint8Array(vec![0; *size as usize]),
            _ => fail!("Go attempted to construct Uint8Array with bad args: {args:?}"),
        },
        DATE_ID => DynamicObject::Date,
```

This had two problems. First, code was grouped by things like "object construction" and "method calls", as opposed to being grouped by "filesystem" and "crypto" objects. Second, this relied heavily on assigning static IDs for everything, which didn't work out well as the system got more complicated. When manually assigning IDs, it's easy to mess up reference counting and such. The new approach separates the implementation of JS methods from the implementation details of the JS engine. Instead of the imperative style of a match statement for every possible method, this PR changes to a declarative style where objects and their methods are created as they would be in JS. This is slightly slower, but much clearer, and it's harder to make mistakes in this model. Here's an example:

```rs
/// in globals, we create the Date constructor
object.insert_func("Date", |_env, _go, _args| Ok(make_date_object()));

// make_date_object is defined as such:
fn make_date_object() -> JsValue {
    let date = JsObject::default();
    date.insert_func("getTimezoneOffset", |_env, _go, _args| {
        Ok(JsValue::Number(0.))
    });
    date.into()
}
```

The JS method implementations in `runtime.rs` never need to care about reference counting or object IDs. They can operate solely on high-level JS values.

### CODE REVIEW

File: arbitrator/jit/src/gostack.rs [hunk 1]
```diff
@@ -296,16 +304,24 @@ impl GoStack {
         let Some(resume) = &env.exports.resume else {
             return Escape::failure(format!("wasmer failed to bind {}", "resume".red()));
         };
+        // recursively call into wasmer (reentrant)
+        resume.call(store)?;
+        self.refresh(env, store)
+    }
+
+    /// Refreshes the stack pointer after potentially the Go runtime was potentially resumed.
```
Comment (tsahee) [inline arbitrator/jit/src/gostack.rs]: potentially *2

File: arbitrator/wasm-libraries/go-js/src/core.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            JsValue::Array(_) => 1,
+            JsValue::String(_) => 2,
+            // Symbols are 3 but we don't support them
+            JsValue::Function(_) => 4,
+        }
+    }
+}
+
+impl PartialEq for JsValue {
+    fn eq(&self, other: &Self) -> bool {
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/core.rs]: Is this SameValueZero (worth a comment)?

File: arbitrator/wasm-libraries/go-js/src/core.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            JsValue::Undefined => write!(f, "undefined"),
+            JsValue::Null => write!(f, "null"),
+            JsValue::Bool(x) => write!(f, "{x}"),
+            JsValue::Number(x) => write!(f, "{x}"),
+            JsValue::String(x) => write!(f, "{x:?}"),
+            JsValue::Object(x) => write!(f, "{x:?}"),
+            JsValue::Uint8Array(x) => write!(f, "{x:?}"),
+            JsValue::Array(x) => write!(f, "{x:?}"),
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/core.rs]: type should also be displayed somehow..

File: arbitrator/wasm-libraries/go-js/src/core.rs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        };
+        let expected_id = JsValueId::new(value.go_typecode(), pool_id);
+        if id.0 != expected_id.0 {
+            panic!("Got non-canonical JS ValueID {id:?} but expected {expected_id:?}");
+        }
+        value.clone()
+    }
+
+    /// Warning: this increments the reference count for the returned id
+    pub fn value_to_id(&self, value: JsValue) -> JsValueId {
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/core.rs]: function name should reflect the fact we add a refcount

File: arbitrator/wasm-libraries/go-js/src/core.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    /// We follow the JS [SameValueZero](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) rule of equality.
+    fn equality(&self) -> JsValueEquality<'_> {
+        match self {
+            JsValue::Undefined => JsValueEquality::AlwaysEqual,
+            JsValue::Null => JsValueEquality::AlwaysEqual,
+            JsValue::Bool(x) => JsValueEquality::Bool(*x),
+            // Treat all NaN values as equal
+            JsValue::Number(x) if x.is_nan() => JsValueEquality::Number(CANONICAL_NAN_BITS),
+            // Treat all zero values as equal
+            JsValue::Number(x) if *x == 0. => JsValueEquality::Number(0_f64.to_bits()),
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/core.rs]: I assume this works for both +0. and -0.? even so.. better to just check against both to make it apparent

**Review by tsahee — COMMENTED** [review]
very initial. didn't go over everything yet.

File: arbitrator/wasm-libraries/go-js/src/core.rs (hunk 2 above)
Comment (tsahee) [reply arbitrator/wasm-libraries/go-js/src/core.rs]: also - maybe a small comment why you chose that

**Comment by rachel-bousfield** [comment]
I've added some additional commits, and think the PR is pretty much good to go. 

Highlights include
- the new EVM Api, which introduces a much safer model for calling Go closures in Rust.
- making call-and-recover an atomic operation, avoiding manual (and error-prone) management of the stack pointer.
- fixing `fs.write` and `undefined.Type()`

File: arbitrator/wasm-libraries/go-js/src/core.rs (hunk 4 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/go-js/src/core.rs]: Renamed to `assign_id` + a comment :)

File: arbitrator/jit/src/gostack.rs [hunk 6]
```diff
@@ -291,36 +290,18 @@ impl GoStack {
...
-        // recover the stack pointer
-        let pointer = get_stack_pointer.call(store)? as u32;
-        self.sp = pointer;
-        self.top = pointer + saved;
-        Ok(())
+        let bytes = self.read_slice(ptr, len);
+        match String::from_utf8(bytes) {
+            Ok(s) => s,
+            Err(e) => {
+                let bytes = e.as_bytes();
```
Comment (tsahee) [inline arbitrator/jit/src/gostack.rs:309]: was that intentional? won't it just print the error twice in the print below?

File: arbitrator/jit/src/syscall.rs [hunk 7]
```diff
@@ -2,667 +2,238 @@
...
-            GoValue::Undefined
-        }
-    }
-}
+use arbutil::Color;
+use go_js::{JsEnv, JsValueId};
+use wasmer::{StoreMut, TypedFunction};
 
 /// go side: λ(v value)
 // TODO: implement ref counting
```
Comment (tsahee) [inline arbitrator/jit/src/syscall.rs]: this TODO is prob. not relevant?

File: arbitrator/wasm-libraries/go-js/src/js_core.rs [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+        };
+        let expected_id = JsValueId::new(value.go_typecode(), pool_id);
+        if id.0 != expected_id.0 {
+            panic!("Got non-canonical JS ValueID {id:?} but expected {expected_id:?}");
+        }
+        value.clone()
+    }
+
+    /// Assigns an id for the given value, incrementing its reference count if already present.
+    pub fn assign_id(&self, value: JsValue) -> JsValueId {
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/js_core.rs:369]: Not mandatory in any way, but wanted to point out the original go-js implementation's version is "storeValue", which gets both the value and an address to store it. There isn't much representation through the code of value ids - only values that are stored/loaded as 8-bytes into/from memory. It has some downside, but makes the code overall less confusing, especially ref counting seems to make sense to me (because you only increase refcount when you actually store this reference)

File: arbitrator/wasm-libraries/go-js/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                panic!("Go attempted to call valueLength on invalid type: {x:?}");
+            }
+        };
+        len
+    }
+
+    pub fn copy_bytes_to_go(&self, src: JsValueId, write_bytes: impl FnOnce(&[u8])) {
+        match self.values.id_to_value(src) {
+            JsValue::Uint8Array(array) => write_bytes(&array.lock()),
+            x => {
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/lib.rs]: should return an error, caller should write 0 (instead of 1) to stack
should not panic

File: arbitrator/wasm-libraries/go-js/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                panic!("Go attempted to call copyBytesToGo on invalid type: {x:?}");
+            }
+        };
+    }
+
+    pub fn copy_bytes_to_js(&self, dest: JsValueId, write_bytes: impl FnOnce(&mut [u8])) {
+        match self.values.id_to_value(dest) {
+            JsValue::Uint8Array(array) => write_bytes(&mut array.lock()),
+            x => {
+                panic!("Go attempted to call copyBytesToJs on invalid type: {x:?}");
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/lib.rs]: should return an error, caller should write 0 (instead of 1) to stack
should not panic

File: arbitrator/wasm-libraries/go-js/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
+// Copyright 2021-2023, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE
+
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/lib.rs:3]: we should point somewhere to the original go.js implementation: https://github.com/golang/go/blob/go1.20.10/misc/wasm/wasm_exec.js
there is a comment pointing there from machine.rs, but it's for a small detail and should reference it for the entire implementation

File: arbitrator/wasm-libraries/go-js/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    pub fn string_val(&self, s: String) -> JsValueId {
+        self.values.assign_id(JsValue::String(Arc::new(s)))
+    }
+
+    pub fn value_length(&self, array: JsValueId) -> usize {
+        let len = match self.values.id_to_value(array) {
+            JsValue::Array(array) => array.lock().len(),
+            JsValue::Uint8Array(array) => array.lock().len(),
+            x => {
```
Comment (tsahee) [inline arbitrator/wasm-libraries/go-js/src/lib.rs]: exists for String
exists for functions
could exist for object
returns "undefined" for bool / number and doesn't panic

File: arbitrator/jit/src/user/evm_api.rs [hunk 13]
```diff
@@ -95,45 +96,13 @@ pub(super) fn exec_wasm(
...
-                    };
-                    outs.push(ApiValue(x));
-                }
-
-                for id in object_ids {
-                    env.js_state.pool.remove(id);
-                }
+                let js_env = &mut WasmerJsEnv::new(sp, &mut store, exports, &mut env.go_state)?;
+                let outs = js_state.call_stylus_func(api_id, method, args, js_env)?;
                 respond.send(outs).unwrap();
```
Comment (tsahee) [inline arbitrator/jit/src/user/evm_api.rs:106]: in js_value_new (below):
is it o.k we're returning nil for error in case an error is returned? maybe better to create some simple error value?

File: contracts
Comment (tsahee) [inline contracts:1]: probably doesn't belong in this PR

File: go-ethereum
Comment (tsahee) [inline go-ethereum:1]: probably doesn't belong in this PR

File: system_tests/program_test.go [hunk 14]
```diff
@@ -40,6 +39,7 @@ import (
 	"github.com/offchainlabs/nitro/validator/valnode"
 	"github.com/wasmerio/wasmer-go/wasmer"
 
+	"github.com/ethereum/go-ethereum/eth/tracers"
```
Comment (tsahee) [inline system_tests/program_test.go]: remove change.. or move tracers/js to be with the rest of go-ethereum imports

**Review by tsahee — COMMENTED** [review]
This is still quite initial review and requires some talking.. specifically about when we want to panic and when we don't

We should also add (at least) two tests:
1. some version of the go-js test with necessary adaptations:
https://go.dev/src/syscall/js/js_test.go
2. a test for reference counting, which will create many objects, have them as fields in other objects/instances in arrays etc, probably pass them between different gofuncs, and have some hook to test we have many references in the middle and no unexpected references after all objects are destroyed,

File: go-ethereum
Comment (rachel-bousfield) [reply go-ethereum:1]: This just bumps us to the latest `stylus/stylus` after a previous merge :)

File: contracts
Comment (rachel-bousfield) [reply contracts:1]: This just bumps us to the latest `stylus-contracts/stylus` after a previous merge :)

File: arbitrator/jit/src/user/evm_api.rs (hunk 13 above)
Comment (rachel-bousfield) [reply arbitrator/jit/src/user/evm_api.rs:106]: Sorry which part / error? Did you mean to add this comment somewhere else?

File: arbitrator/wasm-libraries/go-js/src/lib.rs (hunk 12 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/go-js/src/lib.rs]: Go expects an integer rather than a type, so we should probs panic in our case if the result is not a number. I've updated the panic message to clarify this a bit, as well as added an implementation for `String`.

File: arbitrator/jit/src/gostack.rs (hunk 6 above)
Comment (rachel-bousfield) [reply arbitrator/jit/src/gostack.rs:309]: The second print calls `Debug` which includes extra info, and may change across updates to the stdlib. I've tweaked the message to provide the data in hex either way :)

File: arbitrator/wasm-libraries/go-js/src/lib.rs (hunk 9 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/go-js/src/lib.rs]: Agreed, refactored :)

File: arbitrator/wasm-libraries/go-js/src/lib.rs (hunk 10 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/go-js/src/lib.rs]: Agreed, refactored :)

File: arbitrator/wasm-libraries/go-js/src/lib.rs (hunk 11 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/go-js/src/lib.rs:3]: Added, along with [`js.go`](https://github.com/golang/go/blob/go1.20.10/src/syscall/js/js.go)
